transformations/diverse_paraphrase/transformation.py

```diff
@@ -0,0 +1,151 @@
+import random
+import numpy as np
+import torch
+from random import sample
+from transformers import FSMTForConditionalGeneration, FSMTTokenizer
+
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+
+from transformations.diverse_paraphrase.submod.submodopt import SubmodularOpt
+from transformations.diverse_paraphrase.submod.submodular_funcs import trigger_dips
+
+
+class DiverseParaphrase(SentenceOperation):
+    tasks = [TaskType.TEXT_CLASSIFICATION, TaskType.TEXT_TO_TEXT_GENERATION]
+    languages = ["en"]
+    heavy = True
+
+    def __init__(self, augmenter="dips", num_outputs=4):
+        super().__init__()
+
+        seed = 42
+        random.seed(seed)
+        np.random.seed(seed)
+        torch.manual_seed(seed)
+        torch.use_deterministic_algorithms(True)
+        if torch.cuda.is_available():
+            torch.cuda.manual_seed_all(seed)
+
+        assert augmenter in ["dips", "random", "diverse_beam", "beam"]
```

Comment:
I think it can be useful for people to see the difference in the outputs of these: You can show 1-2 sentences with different "augmenters" by passing it in args in test.json
Here is an example: https://github.com/GEM-benchmark/NL-Augmenter/blob/main/filters/keywords/test.json

Reply:
Sure. The build seems to be failing because of test accuracy evaluation (test.json).
Any suggestions?

Reply:
Reminder. Any suggestions on this?

I had earlier set the variable `heavy=True` as in back_translation code.

Reply:
@aadesh11 @AbinayaM02 any idea of this?

Reply:
@kaustubhdhole : I checked the build failure log. The pytest crashed after it downloaded the necessary libraries (all heavy ones). This looks like an OOM issue on the Github runner, that's my best guess.

The following is the specification of the Github runner,
Hardware specification for Windows and Linux virtual machines:
- 2-core CPU
- 7 GB of RAM memory
- 14 GB of SSD disk space

Reply:
@ashutoshml can you please try the following in your branch (on your machine):
`pytest -s --t=diverse_paraphrase
`
and share the output.

Reply:
```shell
test/test_main.py:105:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test/test_main.py:83: in execute_test_case_for_transformation
    execute_sentence_operation_test_case(transformation, test)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

transformation = <transformations.diverse_paraphrase.transformation.DiverseParaphrase object at 0x7fdf0499c110>
test = {'args': {'augmenter': 'dips'}, 'class': 'DiverseParaphrase', 'inputs': {'sentence': 'Alice in Wonderland is a 2010 Am...0'}, {'sentence': 'Alice in Wonderland is an American live action / animated dark fantasy adventure movie from 2010'}]}

    def execute_sentence_operation_test_case(transformation, test):
        filter_args = test["inputs"]
        outputs = test["outputs"]
        perturbs = transformation.generate(**filter_args)
        for pred_output, output in zip(perturbs, outputs):
>           assert pred_output == output["sentence"], get_assert_message(
                transformation, output["sentence"], pred_output
            )
E           AssertionError: Mis-match in expected and predicted output for DiverseParaphrase transformation:
E              Expected Output: Alice in Wonderland is an American live-action / animated dark fantasy adventure film from the year 2010
E              Predicted Output: Alice in Wonderland is an American live action / animated dark fantasy adventure movie from 2010
E           assert 'Alice in Won...vie from 2010' == 'Alice in Won...the year 2010'
E             - Alice in Wonderland is an American live-action / animated dark fantasy adventure film from the year 2010
E             ?                                        ^                                         ^ ^^      ---------
E             + Alice in Wonderland is an American live action / animated dark fantasy adventure movie from 2010
E             ?                                        ^                                         ^^^ ^

test/test_main.py:26: AssertionError
=============================================================================================================================== warnings summary ===============================================================================================================================
test/test_main.py::test_operation[diverse_paraphrase-length]
  /scratche/prod/ashutosh/miniconda3/envs/nlgenv/lib/python3.7/site-packages/torch/_tensor.py:575: UserWarning: floor_divide is deprecated, and will be removed in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values.
  To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor'). (Triggered internally at  /pytorch/aten/src/ATen/native/BinaryOps.cpp:467.)
    return torch.floor_divide(self, other)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================================================================================================================== short test summary info ============================================================================================================================
FAILED test/test_main.py::test_operation[diverse_paraphrase-length] - AssertionError: Mis-match in expected and predicted output for DiverseParaphrase transformation:
=================================================================================================================== 1 failed, 1 warning in 288.53s (0:04:48) ===================================================================================================================
```

Reply:
Okay, which part is the probabilistic part?

Reply:
There are two ways in which probabilistic outputs may happen:

1. When the candidates are generated using beam search, the argmax operation is resolved non-deterministically in the case of two candidates having the same max-scores. 

2. For `dips`, there is a functional maximization operation that occurs resulting in similar non-determinism. The first selected candidate governs how the rest of the candidates are selected. Since the objective is to maximize the lexical-syntactical diversity of the output candidates.

3. In `random` selection, as the name suggests, random candidates are selected from the beam search, which may or may not produce diverse candidates.